### PR TITLE
Fixes #467 Expose codahale metrics by rest service

### LINKED ISSUE #467: Expose codahale metrics by rest service
So that,

we can get the throughput, jvm status to user.

### PR DESCRIPTION
**Pull request by kkasravi**
Metrics (throughput, meter) are now streamed into the browser using websockets. A few issues remain:

- the socket address is hardcoded. We need to add a REST call asking for the streaming address
- the metrics should be enabled dynamically, also via a REST call.
- the output is only being streamed to the browser console, so we need meter and histogram widgets

I'll open separate issues on the above.

Sample output is below
````
 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task1_0.throughput","count":"0","mean_rate":0,"m1":0,"m5":0,"m15":0,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task1_1.throughput","count":"0","mean_rate":0,"m1":0,"m5":0,"m15":0,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"4824","min":"0","max":"2","mean":0.017509727626459144,"stddev":0.15186226303946004,"median":0,"p75":0,"p95":0,"p98":0,"p99":1,"p999":2}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"4845","min":"0","max":"3","mean":0.02529182879377432,"stddev":0.1906847509548179,"median":0,"p75":0,"p95":0,"p98":1,"p99":1,"p999":3}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"322574","min":"0","max":"257","mean":10.728599221789883,"stddev":15.30848894155429,"median":5,"p75":19,"p95":31.549999999999955,"p98":45.25999999999988,"p99":58.710000000000036,"p999":253.1140000000005}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"745941","min":"0","max":"204","mean":19.92509727626459,"stddev":24.10887412505352,"median":16,"p75":28,"p95":63,"p98":91.83999999999992,"p99":108.71000000000004,"p999":203.1010000000001}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task0_2.throughput","count":"5403540","mean_rate":183232.45704208137,"m1":92708.48408242757,"m5":58156.65737226059,"m15":51516.751634258755,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task0_3.throughput","count":"5427170","mean_rate":184032.84478161833,"m1":91360.3410237763,"m5":55427.11283530134,"m15":48519.20508003388,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task1_2.throughput","count":"0","mean_rate":0,"m1":0,"m5":0,"m15":0,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task1_3.throughput","count":"0","mean_rate":0,"m1":0,"m5":0,"m15":0,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"8571","min":"0","max":"4","mean":0.021400778210116732,"stddev":0.21055216247999536,"median":0,"p75":0,"p95":0,"p98":0,"p99":1,"p999":3.9710000000000036}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"8616","min":"0","max":"1","mean":0.011673151750972763,"stddev":0.10746219270851418,"median":0,"p75":0,"p95":0,"p98":0,"p99":1,"p999":1}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"760347","min":"0","max":"166","mean":10.802529182879377,"stddev":14.733296399933725,"median":5,"p75":19,"p95":31,"p98":50.41999999999996,"p99":66,"p999":164.14400000000023}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"1190110","min":"0","max":"523","mean":18.89591439688716,"stddev":44.85156589065409,"median":10,"p75":23,"p95":47.549999999999955,"p98":171.41999999999996,"p99":208.1300000000001,"p999":523}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task0_0.throughput","count":"9600130","mean_rate":215851.3315306094,"m1":130252.01744908477,"m5":65579.25228108342,"m15":51747.423379528045,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task0_1.throughput","count":"9650650","mean_rate":216985.53109769948,"m1":130455.7266643321,"m5":65457.60155521281,"m15":51545.34079973788,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task1_0.throughput","count":"0","mean_rate":0,"m1":0,"m5":0,"m15":0,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task1_1.throughput","count":"0","mean_rate":0,"m1":0,"m5":0,"m15":0,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"7918","min":"0","max":"2","mean":0.014591439688715954,"stddev":0.135230800596823,"median":0,"p75":0,"p95":0,"p98":0,"p99":1,"p999":2}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"7962","min":"0","max":"1","mean":0.01556420233463035,"stddev":0.1238421458205842,"median":0,"p75":0,"p95":0,"p98":0,"p99":1,"p999":1}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"529107","min":"0","max":"257","mean":10.130350194552529,"stddev":14.20775509512902,"median":3.5,"p75":19,"p95":30,"p98":34.83999999999992,"p99":49.710000000000036,"p999":251.92500000000064}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Histogram",{"count":"1223498","min":"0","max":"173","mean":16.782101167315176,"stddev":19.95796923999278,"median":14,"p75":26,"p95":53,"p98":73.41999999999996,"p99":94.71000000000004,"p999":172.39100000000008}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task0_2.throughput","count":"8869620","mean_rate":199413.65377460243,"m1":124289.781369239,"m5":66808.89289561761,"m15":54559.00978649944,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task0_3.throughput","count":"8917970","mean_rate":200500.24958070845,"m1":123518.07218440622,"m5":64268.78299059287,"m15":51629.79999029154,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task1_2.throughput","count":"0","mean_rate":0,"m1":0,"m5":0,"m15":0,"rate_unit":"events/second"}]
app.js:92 MESSAGE ["org.apache.gearpump.metrics.Metrics.Meter",{"name":"app1.task1_3.throughput","count":"0","mean_rate":0,"m1":0,"m5":0,"m15":0,"rate_unit":"events/second"}]
app.js:89
````

[Review on Reviewable](https://reviewable.io/reviews/intel-hadoop/gearpump/520)

### CODE REVIEW

**Comment by kkasravi** [comment]
This also fixes #287. Below is a data flow diagram for the event bus

screen shot 2015-02-21 at 10 33 20 pm

File: services/dashboard/scripts/app.js [hunk 1]
```diff
@@ -66,4 +66,55 @@ angular.module('app', [
     return page === currentRoute || new RegExp(page).test(currentRoute) ? 'active' : '';
   };
 
+})
+.factory('StreamingService', function() {
```
Comment (stanleyxu2005) [inline services/dashboard/scripts/app.js]: Move StreamingService to a separate file

File: services/dashboard/scripts/app.js [hunk 2]
```diff
@@ -66,4 +66,55 @@ angular.module('app', [
...
+    StreamingService.prototype.send = function (data) {
+      function send() {
+        console.log("connected="+self.websocket.readyState);
+        switch(self.websocket.readyState) {
+          case 1:
+            console.log("SENDING "+data);
+            self.websocket.send(data);
+            break;
+          default:
+            setTimeout(send, 500);
```
Comment (stanleyxu2005) [inline services/dashboard/scripts/app.js]: Better use a global configuration object to hold the options.
```
angular.constant('CONF', {WebSocketSendTimeout:500});
```

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala [hunk 3]
```diff
@@ -176,7 +176,7 @@ class TaskActor(val taskContextData : TaskContextData, userConf : UserConfig, va
         msg match {
           case SendAck(ack, targetTask) =>
             transport(ack, targetTask)
-            LOG.debug("Sending ack back, taget taskId: " + taskId + ", my task: " + taskId + ", received message: " + ack.actualReceivedNum)
+            LOG.info("Sending ack back, taget taskId: " + taskId + ", my task: " + taskId + ", received message: " + ack.actualReceivedNum)
```
Comment (stanleyxu2005) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala]: Logging in general: Lazy string evaluation will bring performance advantage?
```
LOG.info("Sending ack back. taskId={}, receivedMessages: {}", taskId, ack.actualReceivedNum);
```

File: core/src/main/scala/org/apache/gearpump/metrics/Metrics.scala [hunk 4]
```diff
@@ -48,6 +51,15 @@ object Metrics extends ExtensionId[Metrics] with ExtensionIdProvider {
   val LOG: Logger = LogUtil.getLogger(getClass)
   import org.apache.gearpump.util.Constants._
 
+  sealed trait MetricType
+  case class Histogram(count: Long, min: Long, max: Long, mean: Double, stddev: Double, median: Double, p75: Double, p95: Double, p98: Double, p99: Double, p999: Double) extends MetricType
+  case class Counter(name: String, value: Long) extends MetricType
+  case class Meter(name: String, count: Long, mean_rate: Double, m1: Double, m5: Double, m15: Double, rate_unit: String) extends MetricType
+  case class Timer(name: String, count: Long, min: Double, max: Double, mean: Double, stddev: Double,
+                   median: Double, p75: Double, p95: Double, p98: Double, p99: Double, p999: Double, mean_rate: Double,
+                   m1: Double, m5: Double, m15: Double, rate_unit: String, duration_unit: String) extends MetricType
```
Comment (stanleyxu2005) [inline core/src/main/scala/org/apache/gearpump/metrics/Metrics.scala]: Most APIs uses camelcase for parameters, but rate_unit and duration_unit here are not.

File: core/src/main/scala/org/apache/gearpump/metrics/Metrics.scala [hunk 5]
```diff
@@ -48,6 +51,15 @@ object Metrics extends ExtensionId[Metrics] with ExtensionIdProvider {
   val LOG: Logger = LogUtil.getLogger(getClass)
   import org.apache.gearpump.util.Constants._
 
+  sealed trait MetricType
+  case class Histogram(count: Long, min: Long, max: Long, mean: Double, stddev: Double, median: Double, p75: Double, p95: Double, p98: Double, p99: Double, p999: Double) extends MetricType
```
Comment (stanleyxu2005) [inline core/src/main/scala/org/apache/gearpump/metrics/Metrics.scala]: `Long` type (count, min, max) is actually interpreted as string in output.

File: services/dashboard/scripts/app.js (hunk 1 above)
Comment (kkasravi) [inline services/dashboard/scripts/app.js]: I've defined a services.js which adds the the 'app' module.

File: services/dashboard/scripts/app.js (hunk 2 above)
Comment (kkasravi) [inline services/dashboard/scripts/app.js]: Added this in services.js prior to the StreamingService declaration.

File: streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala (hunk 3 above)
Comment (kkasravi) [inline streaming/src/main/scala/org/apache/gearpump/streaming/task/TaskActor.scala]: Correct - a better form is            
```LOG.info(s"Sending ack back, target taskId: $taskId, my task: $taskId, received message: ${ack.actualReceivedNum}")```

File: core/src/main/scala/org/apache/gearpump/metrics/Metrics.scala (hunk 4 above)
Comment (kkasravi) [inline core/src/main/scala/org/apache/gearpump/metrics/Metrics.scala]: Corrected.

File: core/src/main/scala/org/apache/gearpump/metrics/Metrics.scala (hunk 5 above)
Comment (kkasravi) [inline core/src/main/scala/org/apache/gearpump/metrics/Metrics.scala]: We may need javascript class's which uPickle provides. Let's handle in subsequent task.

**Comment by kkasravi** [comment]
Updated with Stanley's changes integrated. Thanks Stanley for good feedback.

**Comment by kkasravi** [comment]
Sorry - I merged to master w/o any +1 from members including Sean. Please revert if you have concerns. I had meant just to update the PR with Stanley's suggested changes. Thanks Kam.

**Pull request merged**